include/dr/shp/algorithms/transform.hpp

```diff
@@ -0,0 +1,55 @@
+// SPDX-FileCopyrightText: Intel Corporation
+//
+// SPDX-License-Identifier: BSD-3-Clause
+#pragma once
+
+#include <dr/shp/init.hpp>
+#include <dr/shp/util.hpp>
+
+namespace shp {
+
+namespace __detail {
+template <class T> T *get_local_pointer(shp::device_ptr<T> r) {
+  return r.local();
+}
+template <class T> T *get_local_pointer(T *t) { return t; }
+} // namespace __detail
+
+template <class ExecutionPolicy>
+void transform(ExecutionPolicy &&policy, lib::distributed_range auto &&in,
+               lib::distributed_iterator auto out, auto && fn) {
+
+  static_assert( // currently only one policy supported
+      std::is_same_v<std::remove_cvref_t<ExecutionPolicy>, device_policy>);
+
+  std::vector<sycl::event> events;
+  using OutT = typename decltype(out)::value_type;
+  std::vector<device_vector<OutT, device_allocator<OutT>>*> buffers;
+
+  for (auto&& [in_seg, out_seg] : views::zip(in, rng::subrange(out, out+rng::size(in))).zipped_segments()) {
+
+    sycl::queue q(shp::context(), policy.get_devices()[in_seg.rank()]);
+    const std::size_t seg_size = rng::size(in_seg);
+    assert(seg_size == rng::size(out_seg));
+    auto local_in_seg = __detail::get_local_segment(in_seg);
+
+    if (in_seg.rank() == out_seg.rank()) {
+      auto local_out_seg = __detail::get_local_segment(out_seg);
+      events.emplace_back(
+          q.parallel_for(seg_size, [=](auto idx) { local_out_seg[idx] = fn(local_in_seg[idx]); }));
```

Comment:
You are doing a lambda capture of something that has pointers? I don't think that will work.

Reply:
I wrote it this way and found problems with using `shp::for_each`. I was loosing reference when zipping/unzipping `in/out` and `out` was finally not modified. I gave up being not sure if when I finally will make it working it will became more readable and as fast a current version, which may be easier for compiler to optimize.

But I don't object if anyone wants to try to use `shp::for_each` with this code - all UTs are already here, pass in orlab, so refactoring should be safe.